### ISSUE TITLE
String 'dms' was not recognized as a valid DateTime.

### ISSUE THREAD

**Issue by veronika-lammon** [thomasloupe/Slackord#72]
**Slackord version (please complete the following information):**
 - OS: Mac
 - Version: 2.4.10

**Describe the bug**
when trying to import json data getting error: String 'dms' was not recognized as a valid DateTime.
drwxrwxr-x@  8 vbronnik  staff   256B Apr 21 20:38 ../
-rw-r--r--@  1 vbronnik  staff   6.0K Apr 21 20:35 .DS_Store
drwxrwxr-x   3 vbronnik  staff    96B Nov 30  1979 D03J7C7S2Q7/
drwxrwxr-x   9 vbronnik  staff   288B Nov 30  1979 D03KBNW0TLG/
-rw-rw-r--@  1 vbronnik  staff   626B Apr 21 19:09 dms.json
drwxrwxr-x   3 vbronnik  staff    96B Nov 30  1979 education/
drwxrwxr-x   6 vbronnik  staff   192B Nov 30  1979 fun/
drwxrwxr-x  14 vbronnik  staff   448B Nov 30  1979 general/
-rw-rw-r--   1 vbronnik  staff     3B Apr 21 19:09 groups.json
drwxrwxr-x  16 vbronnik  staff   512B Nov 30  1979 homework/
drwxrwxr-x   3 vbronnik  staff    96B Nov 30  1979 marketing/
drwxrwxr-x   5 vbronnik  staff   160B Nov 30  1979 meetings/
-rw-rw-r--   1 vbronnik  staff     3B Apr 21 19:09 mpims.json
drwxrwxr-x   3 vbronnik  staff    96B Nov 30  1979 random/
-rw-rw-r--   1 vbronnik  staff    10K Apr 21 19:09 users.json
drwxrwxr-x   5 vbronnik  staff   160B Nov 30  1979 website/
drwxrwxr-x   3 vbronnik  staff    96B Nov 30  1979 website-copy/
drwxrwxr-x   3 vbronnik  staff    96B Nov 30  1979 website-design/
drwxrwxr-x   4 vbronnik  staff   128B Nov 30  1979 website-media/

Please select an option:
1. Import Slack channels to Discord.
2. Import Slack messages.

Selection: 2
Not a valid option...
String 'dms' was not recognized as a valid DateTime.

Please select an option:
1. Import Slack channels to Discord.
2. Import Slack messages.

**To Reproduce**
1. Select JSON folder
2. Type /slackord
3. See error

**Screenshots/Videos, etc**

**Additional context**

**Comment by thomasloupe** [comment]
Hello @veronika-lammon. Welcome!

Please remove the following files from your directory:
`dms.json` and `groups.json`, then try again.

**Comment by veronika-lammon** [comment]
Hello @thomasloupe 
It looks like the same error is showing up for all the .json files. As soon as I remove dms, the next one complains, then remove that one, the third one complains...and so on.

**Comment by thomasloupe** [comment]
So the way you should be placing files into the `Files` directory is like this:

`Files`
`-File1.json`
`-File2.json`
`-File3.json`

There should be no folders inside the `Files` directory.

**Comment by veronika-lammon** [comment]
Where would I put all the folders which contain all the channel data? Should those be put under the main Slackord folder?
Also, where would the user folder live too?

Identity:
drwxrwxr-x 3 vbronnik staff 96B Nov 30 1979 D03J7C7S2Q7/
drwxrwxr-x 9 vbronnik staff 288B Nov 30 1979 D03KBNW0TLG/

Channel data:
drwxrwxr-x 3 vbronnik staff 96B Nov 30 1979 education/
drwxrwxr-x 6 vbronnik staff 192B Nov 30 1979 fun/
drwxrwxr-x 14 vbronnik staff 448B Nov 30 1979 general/
drwxrwxr-x 16 vbronnik staff 512B Nov 30 1979 homework/
drwxrwxr-x 3 vbronnik staff 96B Nov 30 1979 marketing/
drwxrwxr-x 5 vbronnik staff 160B Nov 30 1979 meetings/
drwxrwxr-x 3 vbronnik staff 96B Nov 30 1979 random/
drwxrwxr-x 5 vbronnik staff 160B Nov 30 1979 website/
drwxrwxr-x 3 vbronnik staff 96B Nov 30 1979 website-copy/
drwxrwxr-x 3 vbronnik staff 96B Nov 30 1979 website-design/
drwxrwxr-x 4 vbronnik staff 128B Nov 30 1979 website-media/

Thank you
V

**Comment by thomasloupe** [comment]
You place a single folder's contents inside of the `Files` folder. Leave the other folders outside of the `Files` directory for now. Just choose one, place the files from inside the folder into the `Files` folder.

`channels.json` should be placed in the root directory of Slackord as shown in the README.

**Comment by veronika-lammon** [comment]
Yeah, the channels.json at the top level I was able to use to create the channels but it seems there is no "batch" way to upload attachments and migrate messages for all channels currently. Is that correct?

**Comment by thomasloupe** [comment]
That's correct, though I am planning at some point to remove the process of creating channels separately, and instead just allow you to place all your folders inside the `Files` directory, and it will auto-create the channels and post the contents to each one. Will just take time to get that feature implemented.

**Comment by veronika-lammon** [comment]
Is there a special way to allow for attachments to be uploaded currently?

**Comment by thomasloupe** [comment]
Can you give specific filetype extensions that are missing from the import?
Does it appear to be something like an executable, or some other kind of filetype extension?

Images and basic file attachments should always be attached, but there may be filetypes that I haven't accounted for yet.

**Comment by veronika-lammon** [comment]
Slackdump saves off attachments in the channel, so like jpgs and such. so they are included in the channel name folder

Slackdump.zip -> <name of the channel> -> attachments -> attachment-file.jpg

**Comment by thomasloupe** [comment]
Ah, yeah, Slackdump exports are compatible with Slackord, but the problem is that Slackord has no way of really knowing whether or not you're using Slack exports or Slackdump exports, so it has no idea of the attachments zip at all. It only looks at the JSON data in each message that is posted. If it's not in the JSON, then it doesn't really exist to Slackord.

**Comment by rusq** [comment]
Hey @thomasloupe, if you're interested, and if there's a way to import attachments into discord, we could potentially collaborate on that.

Slackdump supports the "mattermost" style of attachment storage, and I think it's the easiest way to programmatically access files, and detect if they are included in the archive.

The mattermost type export directory structure looks exactly like the traditional slack export, with an addition of a single `__uploads` directory.  This directory contains all attachments from all channels in one place.

The attachment storage structure is the following:

```
_uploads/<FILE_ID>/<filename.ext>
```

This is the listing of an example archive exported by slackdump into mattermost export type archive:
```
$ unzip -l my_export.zip

Archive:  my_export.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
...
        0  00-00-1980 00:00   random/
    12628  05-04-2023 16:43   random/2023-04-24.json
     3723  05-04-2023 16:43   random/2023-04-29.json
        0  00-00-1980 00:00   __uploads/
        0  00-00-1980 00:00   __uploads/F055C6558RL/
    74880  05-04-2023 16:43   __uploads/F055C6558RL/starseeker.jpg
        0  00-00-1980 00:00   DHYNUJ00Y/
     2182  05-04-2023 16:43   DHYNUJ00Y/2019-05-22.json
     4962  05-04-2023 16:43   DHYNUJ00Y/2019-07-25.json
...
     5017  05-04-2023 16:43   channels.json
        3  05-04-2023 16:43   groups.json
        3  05-04-2023 16:43   mpims.json
      848  05-04-2023 16:43   dms.json
    14742  05-04-2023 16:43   users.json
---------                     -------
   170531                     32 files
```

The message JSON, that links to that file, is the following:
```jsonc
[
  {
    "client_msg_id": "db462974-8319-4f4d-918f-739b9b3315db",
    "type": "message",
    "user": "UHSD97ZA5",
    "text": "Test file upload",
    "ts": "1682743815.053209",
    "thread_ts": "1682335799.257359",
    "parent_user_id": "UHSD97ZA5",
    "files": [
      {
        "id": "F055C6558RL",
//...
        "name": "starseeker.jpg",
        "title": "starseeker.jpg",
//...
      }
    ],
    "replace_original": false,
//...
  }
]
```

I left only significant fields of the file structure above.


Let me know what you think?

**Comment by thomasloupe** [comment]
Hey @rusq!

If the entire export is exactly the same as slack, then it shouldn't be super difficult to upload attachments.

There are a couple small boundaries to cross first but I'm definitely open to this.

There's only one roadblock I see - currently, Slackord has no way of really knowing if an export is a Slack export or a Slackdump export, but probably, and this is off the top of my head, the best route would be to place some kind of identifier key in the JSON somewhere so I know to use the uploads directory. Otherwise, I think we'd just have to rely on seeing the uploads folder inside of the directory, ignore it for parsing messages, but obviously use it for file attachments.

It's definitely possible either way though.

I will need to work through the next major update and then I think we can visit this, though I don't have an ETA for the update just yet. 

I'd encourage you to join the community discord, as there are lots of folks using Slackdump that sometimes encounter issues with Slackord that I can't really answer, but I'm always happy to do the best I can providing support for it!

**Comment by thomasloupe** [comment]
@rusq Slackord 3.0.0 has made it to release, and assuming I haven't broken everything, I'm going to give it a couple weeks, then will reach out about this.

Thanks again!

**Comment by thomasloupe** [comment]
This bug has been resolved in 3.0.3. If there aren't messages in the export, it'll ignore the json file and won't parse it.

@rusq, thanks for staying patient with me for a while on this. The majority of major bugs have been fixed in the upcoming 3.0.3 release. Could you please open a new enhancement issue for getting onboarded with Slackdump so we can track that separately?
